pgjdbc/src/main/java/org/postgresql/util/LruCache.java

```diff
@@ -36,14 +36,16 @@
     Value create(Key key) throws SQLException;
   }
 
-  private final @Nullable EvictAction<Value> onEvict;
-  private final @Nullable CreateAction<Key, Value> createAction;
-  private final int maxSizeEntries;
-  private final long maxSizeBytes;
-  private long currentSize;
+  private final EvictAction<Value> onEvict;
+  private final CreateAction<Key, Value> createAction;
   private final Map<Key, Value> cache;
+  final int maxSizeEntries;
+  final long maxSizeBytes;
```

Comment:
Why is it necessary for these values to be exposed to the package scope? From an API perspective, it provides maximum flexibility to not expose private names, as it avoids people accidentally relying on them.

Reply:
They are referenced from the LimitedMap. If private, this requires synthetic accessors to be generated (at default scope). By changing to default scope, they can be directly accessed.
At default scope, no consumer has visibility, so they cannot rely on them.

Reply:
If they need to be accessed outside of the class, perhaps they should be real accessors? Otherwise, it represents an action at a distance, as well as potential mutability concerns as other classes in the same package can read and write these field members (although the "final" should minimize this impact, and have these fields be treated as immutable).

The logic around avoiding synthetic accessors, could be used to avoid many higher level language features. For example, lambda also has overhead, involving various synthesized methods and classes, but we still use them, because it makes the code easier to read.

I know some tools (SonarQube?) do try to suggest issues such as these, but I'm wary of them making design decisions here. Either these fields should be hidden, in which case "private" makes sense (and we shouldn't care how it gets implemented, as long as it isn't demonstrated to be a problem), or they should be exposed read-only, in which case a "get" accessor with the right scope, makes sense. The option of using a package-visible field member, is an option to expose internal API outside of this file, and allow it to be accidentally made use of in the future, which then makes it part of the API. You could similarly argue why not always use "package" scope for all field members. That it might be used in a synthetic method doesn't really change the logic of whether you should or shouldn't do it for a particular use case.

"Synthetic accessors" in this case, is a matter of convenience for a particular use case of inner classes. However, having an accessor is not necessarily the problem. Having an *implicit* accessor might be the problem, although I wouldn't over-rotate on it too much. If you need internal state from this class to be directly accessed by an inner class, then synthetic accessors are a built-in feature that works fine. I doubt you would be able to prove a performance issue. But, if you moved the inner class to a separate class file, possibly even a different package, it would force a review of the API to ensure the API was right, and this is the point where you would be forced to decide whether it is private or not, and how to protect concurrent access. If it should be be exposed, using an accessor method allows the interface to be preserved, even if they underlying implementation changes.

Java has a sophisticated JIT. Some of the theory around what is expensive and what is not expensive, is without awareness of the JIT. As a particular example here, the JIT might eliminate the synthetic accessor from the hot runtime paths, permitting inline and direct access, meaning that the second-guessing of making it package-scoped to avoid the synthetic accessor, was also pre-mature optimization but with a side effect that now internal fields are visible outside of this class.

Reply:
I'll revert the visibility changes from this PR and include in another.

Ultimately I do not think we are that far apart other than the impact of moving from private to package scope. While I certainly prefer to use private as much as possible, I am willing to change to default scope pretty quickly for things like avoiding compiler warnings, easier unit testing, etc. From an API perspective, I still view default scope as highly encapsulated and easily controlled/regulated. I am far more circumspect about changing to protected or public as both of those then give access to those outside the project.